server/parsing_lib/log_parser.py

```diff
@@ -8,61 +8,128 @@
+            .get('filterLineRegex', None)
+        exception_regex = LogParser.parser_info[os_type]\
+            .get('exceptionRegex', None)
 
-        # Ensure that the first log is not a "beginning of /dev/log" line
-        while re.search('beginning of', raw_data[0]) is not None:
-            raw_data = raw_data[1:]
-
-        # Parse the first log line to have context for futher log lines if an
-        # event was split across multiple lines.
-        old_log = LogParser.parse_raw_log(raw_data[0])
-        log_entries.append(LogParser.parse_entries(old_log))
         current_log = None
+        in_unhandled_exception = False
+        multiline = False
 
-        # Since we've already parsed the first line, start at index 1.
-        for line in raw_data[1:]:
-            # Skip "beginning of /dev/log" lines. There may be cases when these
-            # appear in a log line that is not at the beginning of the raw data
-            if re.search('beginning of', line) is not None:
+        for line in raw_log_lines.splitlines():
+            # Skip lines that are not log lines. There may be cases when these
+            # appear in a log line that is not at the beginning of the raw
+            # data.
+            if filter_regex and filter_regex.match(line):
                 continue
 
-            # check if current log is like the previous log parsed
-            current_log = LogParser.parse_raw_log(line)
-            if current_log['time'] != old_log['time']:
-                log_entries.append(LogParser.parse_entries(current_log))
+            # Check to see if an iOS unhandled exception is starting.
+            if exception_regex:
+                exception_groups = exception_regex.match(line)
+                if exception_groups:
+                    in_unhandled_exception = True
+                    multiline = True
+                    exception_time_string = exception_groups.group(1)
+                    current_log = {
+                        'time': LogParser._parse_datetime(exception_time_string,
+                            os_type),
+                        'logType': 'Error',
+                        'tag': '',
+                        'text': line,
+                    }
+                    continue
+
+            # If we are in an unhandled exception, just add the line to the
+            # current log.
+            if in_unhandled_exception:
+                current_log['text'] += '\n%s' % line
+                multiline = True
             else:
-                # if part of the same event, add the log's text to the previous
-                # parsed log
-                log_entries[-1]['text'] += ('\n%s' % current_log['text'])
-            old_log = current_log
+                # Check if current log has the same time as the previous log
+                # parsed.
+                new_log = LogParser.parse_raw_log(line, os_type)
+                if not current_log or current_log['time'] != new_log['time']:
```

Comment:
Super nitt: I think this if statement makes more sense if you inverse it, so flip the if/else code blocks and change the if-statement to:  `if current_log and current_log['time'] == new_log['time']`